doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+to other optimizations so that other optimizations can take advantage of what inlining has been done and optimize
+the code better.
+
+# 2. The Downside Of Inlining
+
+The downside is that the method gets much bigger if many callees are inlined into a method. In general, for a
+language like Java that has so many calls and is object oriented, it is always beneficial to inline a method on
+the hot path, unless the callee is really large and it executes lots of instructions. A small method on a hot
+path tends to get inlined in a language like Java because of so much calling and object orientedness built into it.
+
+# 3. Balance Multiple Paths Of Optimizations
+
+There is a division of responsibilities among the paths of optimizations in different parts of the JIT. We do not
+want to pollute every optimization with every other optimization. For example, we do not want to do folding or
+dead tree elimination while we are doing inlining. The optimizations that generate the dead trees could have
+added some code to eliminate the dead trees, but we choose not to do that. Instead, the dead trees will be cleaned
+up by dead tree elimination optimization.
+
+When inlining is taking place, we are aware of the opportunity for other optimizations because of inlining. Such
+as when looking at the arguments of the call at the call site and there is a constant `String`, that should increase
+the priority of inlining the call. It makes it more likely to inline the call because there might be other
+optimizations that can happen downstream because of passing in this constant `String`. But inlining will not do
+that optimization, that optimization will be done separately by other optimization paths when they run.
+
+Generally speaking, other optimization do not differentiate if the trees are inlined calls or not.
+
+Occasionally, some optimizations might be interested in where the code comes from. For example, if we know a method
+is from `String` class, we know all the access of the method will be within the array bound. It is impossible to
+raise an exception there. The bound check can be skipped for the method that comes from String class. Such thing
+is normally restricted to sets of Java class library methods. Even if the `String` method is inlined into some
+other arbitrary method `foo()`, that property should not change. The bound check should not be left in place
+simply because `foo()` is being compiled. If it comes from the `String` code, regardless of whether it is inlined
+or not, we should still be able to omit that bound check. This is an example of an awareness that the code comes
+from an inlined method and where we differentiate the inlined code.
+
+There are other enhancements that have been done over the years, such as method handle inlining. There is code
+that analyzes the callee’s bytecode and looks at whether there are opportunities to do some optimization. More
```

Comment:
"More and more code has been added in Inliner to be incrementally aware of the opportunities when going into the callees." -> "Better techniques continue to be added to the inliner (such as argument type propagation and abstract interpretation) in order to determine the target callee to be inlined"